docs/foreign_code.md

```diff
@@ -0,0 +1,156 @@
+
+We hope that the region abstraction will make programming with sandboxes natural within the language.
+
+### Foreign libraries are classes
+
+If the unit of interoperability is a library, how do we surface that within Verona?
+We aim to treat foreign libraries just like any other module: as classes that can contain nested types.
+For example, we may import LLVM in something like this (straw-man syntax):
+
+```verona
+class LLVM
+{
+	[#native, #interface_language="c++"]
+	imports library("libLLVM.so", "llvm.h");
+}
+```
+
+This includes two attributes that tell the compiler the format of the library (native code) and how to parse the header (which, in this example, would have to be something provided by the user that included the desired headers).
+Once C++ module support is more widespread, module imports can take the place of header inclusions.
+
+This instructs the compiler to take the C++ namespace structure and inject each namespace as a class within the LLVM class and all C++ classes as nested classes.
+
+Instantiating this class would require the user to specify a sandboxing policy, for example (again, straw-man syntax):
+
+```verona
+var llvm = LLVM.Create(PROCESS);
+```
+
+This would create a new process that loaded the LLVM library and had RPC invocations for any of the LLVM functions that this program called.
+The `llvm` object in this example is a new isolated region.
+The programmer can then construct LLVM types within that region.
+
+Implementation
+--------------
+
+The proposed implementation strategy for C/C++ interoperability is to deeply embed clang.
+We are already depending on LLVM (and MLIR) for our compiler infrastructure and so embedding clang is not a significant addition.
+In the above example, the we would pass the `llvm.h` header to clang and generate an AST.
+Any lookup of symbols within the `LLVM` class's namespace would inspect this AST to find equivalent types.
+
+We want to avoid having to deal with the C/C++ ABI details in the Verona compiler.
+A suggestion from Chris Lattner provides a path to avoiding that by having clang synthesise wrapper functions.
+
+For example, consider some Verona code that wants to call `llvm::IRBuilder<>::CreateRetVoid`.
+This is a method in a templated class, so the template will need to be instantiated before it can be called and the caller then needs to be aware of C++ ABI details.
+This, like every other problem in computer science, can be addressed by a level of indirection, using clang to synthesise a function roughly equivalent to this :
+
+```c++
+__attribute__((used,always_inline))
+extern "C" llvm::ReturnInst *irbuilder_createretvoid(llvm::IRBuilder<> *b)
+{
+        return b->CreateRetVoid();
+}
+```
+
+This takes and returns all values as pointers and so the LLVM lowering is predictable as a function that takes and returns only pointers.
+For complex return values, the caller will need to allocate space and pass a pointer to the return.
```

Comment:
These will need to be special Verona containers with C++ objects inside, because we want to control their life-time inside Verona, where they will be used, and because the C++ side cannot see them.

If we return a pointer allocated by the C++ code, how do we make sure it isn't also freed by the C++ code before its lifetime expires in Verona?

Reply:
This should be handled by value types already.  Verona value types can live only on the stack and have well-defined lifetimes (module @sylvanc actually extracting that bit of the spec from his brain and writing it on paper).

Reply:
```suggestion
For complex return values, the caller will need to allocate space and pass a pointer to the return.
These will be exposed in Verona as value types, which cannot be stored on the heap in Verona.
```